cibuildwheel/linux.py

```diff
@@ -51,130 +67,132 @@ def build(project_dir, output_dir, test_command, test_requires, test_extras, bef
-            test_command=shlex_quote(
-                prepare_command(test_command, project='/project') if test_command else ''
-            ),
-            before_build=shlex_quote(
-                prepare_command(before_build, project='/project') if before_build else ''
-            ),
-            build_verbosity_flag=' '.join(get_build_verbosity_extra_flags(build_verbosity)),
-            repair_command=shlex_quote(
-                prepare_command(repair_command, wheel='"$1"', dest_dir='/tmp/repaired_wheels') if repair_command else ''
-            ),
-            environment_exports='\n'.join(environment.as_shell_commands()),
-            uid=os.getuid(),
-            gid=os.getgid(),
-        )
-
-        def run_docker(command, stdin_str=None):
-            print('docker command: docker {}'.format(' '.join(map(shlex_quote, command))))
-            if stdin_str is None:
-                subprocess.check_call(['docker'] + command)
-            else:
-                args = ['docker'] + command
-                process = subprocess.Popen(args, stdin=subprocess.PIPE, universal_newlines=True)
-                try:
-                    process.communicate(stdin_str)
-                except KeyboardInterrupt:
-                    process.kill()
-                    process.wait()
-                if process.returncode != 0:
-                    raise subprocess.CalledProcessError(process.returncode, args)
-
         container_name = 'cibuildwheel-{}'.format(uuid.uuid4())
         try:
             run_docker(['create',
                         '--env', 'CIBUILDWHEEL',
                         '--name', container_name,
                         '-i',
-                        '-v', '/:/host', # ignored on Circle
+                        '-v', '/:/host', # ignored on CircleCI
                         docker_image, '/bin/bash'])
             run_docker(['cp', os.path.abspath(project_dir) + '/.', container_name + ':/project'])
-            run_docker(['start', '-i', '-a', container_name], stdin_str=bash_script)
+
+            for config in platform_configs:
+                if dependency_constraints:
+                    constraints_file = dependency_constraints.get_for_python_version(config.version)
+                    run_docker(['cp', os.path.abspath(constraints_file), container_name + ':/constraints.txt'])
+
+                run_docker(['start', '-i', '-a', container_name], stdin_str='''
+                    set -o errexit
+                    set -o xtrace
+                    mkdir -p /output
+                    cd /project
+
+                    PYBIN="{config_python_bin}"
+                    export PATH="$PYBIN:$PATH"
+
+                    # check the active python and pip are in PYBIN
+                    # if `test` returns false, the script will exit due to errexit
+                    test "$(which pip)" = "$PYBIN/pip"
+                    test "$(which python)" = "$PYBIN/python"
+
+                    {environment_exports}
```

Comment:
Cfr. #264, this should probably come before those two test statements?

Reply:
good point 👍